benchmarks/distributed/tensor/tp_benchmark.py

```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/env python3
+#
+# Measure distributed training iteration time.
+#
+# This program performs a sweep over a) a number of model architectures, and
+# b) an increasing number of processes. This produces a 1-GPU baseline,
+# an 8-GPU baseline (if applicable), as well as measurements for however
+# many processes can participate in training.
+#
+
+import argparse
+import itertools
+import json
+import os
+import shlex
+import subprocess
+import sys
+import time
+
+import numpy as np
+import torch
+import torch.distributed as dist
+import torch.nn as nn
+import torch.optim as optim
+import torch.multiprocessing as mp
+from torch.distributed._tensor import (
+    DeviceMesh,
+)
+import contextlib
+from torch.distributed.tensor.parallel import (
+    PairwiseParallel,
+    SequenceParallel,
+    parallelize_module,
+)
+from torch.distributed.nn.functional import (
+    _reduce_scatter_base,
+    _all_gather_base,
+    all_reduce,
+)
+from torch.utils._pytree import tree_flatten, tree_map
+
+def setup(rank, world_size):
+    os.environ['MASTER_ADDR'] = 'localhost'
+    os.environ['MASTER_PORT'] = '12355'
+
+    # initialize the process group
+    dist.init_process_group("nccl", rank=rank, world_size=world_size)
+    torch.cuda.set_device(rank)
+
+def cleanup():
+    dist.destroy_process_group()
+
+
+class DummyDTensor(torch.Tensor):
+    _local_tensor: torch.Tensor
+
+    @staticmethod
+    def __new__(
+        cls,
+        local_tensor: torch.Tensor,
+    ) -> "DummyDTensor":
+        r = torch.Tensor._make_wrapper_subclass(  # type: ignore[attr-defined]
+            cls,
+            local_tensor.size(),
+            strides=local_tensor.stride(),
+            dtype=local_tensor.dtype,
+            device=local_tensor.device,
+            layout=local_tensor.layout,
+            requires_grad=local_tensor.requires_grad,
+        )
+        r._local_tensor = local_tensor.detach()
+        return r
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
+        def unwrap_schema(e: object) -> object:
+            return e._local_tensor if isinstance(e, DummyDTensor) else e
+        args_unwrap = tree_map(unwrap_schema, args)
+        kwargs_unwrap = tree_map(unwrap_schema, kwargs)
+        return func(*args_unwrap, **kwargs_unwrap)
```

Comment:
not only you need to unwrap it, you also need to rewrap it by calling another tree_map to convert the output to dummyDTensor